kolibri/plugins/setup_wizard/middleware.py

```diff
@@ -0,0 +1,20 @@
+from django.core.urlresolvers import reverse
+from django.shortcuts import redirect
+from kolibri.auth.models import DeviceOwner
+
+
+class SetupWizardMiddleware():
+    """
+    display the superuser creation app if no superuser exists.
+    """
+    has_superuser = False
+
+    def process_request(self, request):
+        if request.path.startswith(reverse("facility-list")) or request.path.startswith(reverse("deviceowner-list")):
```

Comment:
To make this faster, start with checking for the cached boolean (since that's the most common case, and is fast to check). You can also shorten it using some boolean logic up front. Something like:

```python
# If a DeviceOwner has already been created, no need to do anything here
self.deviceowner_already_created = self.deviceowner_already_created or DeviceOwner.objects.exists()
if self.deviceowner_already_created:
    return

# Don't redirect for URLs that are required for the setup wizard
allowed_paths = [reverse(name) for name in ["facility-list", "deviceowner-list", "kolibri:setupwizardplugin:setupwizard"]]
if any(request.path.startswith(path_prefix) for path_prefix in allowed_paths):
    return

# If we've gotten this far, we want to redirect to the setup wizard
return redirect(reverse("kolibri:setupwizardplugin:setupwizard"))
```